ykrt/src/mt.rs

```diff
@@ -219,215 +217,206 @@ impl MTThread {
-                    // couldn't have been tracing this location, at which point we can bail out.
-                    // FIXME: this is a terrible spinlock.
-                    loop {
-                        match lp.phase() {
-                            PHASE_TRACING => break,
-                            PHASE_TRACING_LOCK => (),
-                            PHASE_COMPILED | PHASE_COMPILING | PHASE_LOCKED | PHASE_COUNTING
-                            | PHASE_DONT_TRACE => {
-                                // If we observe any of these states, then another thread took
-                                // responsibility for this Location, proving that this thread can't
-                                // be responsible for it.
+                    debug_assert!(!ls.is_locked());
+                    match loc.compare_exchange_weak(
+                        ls,
+                        new_ls,
+                        Ordering::Acquire,
+                        Ordering::Relaxed,
+                    ) {
+                        Ok(_) => {
+                            // We've initialised this Location and obtained the lock, so we can now
+                            // start tracing for real.
+                            let tid = Arc::clone(&self.inner.tid);
+                            let tt = start_tracing(self.inner.tracing_kind);
+                            *unsafe {new_ls.hot_location()} = HotLocation::Tracing(Some((tid, tt)));
+                            Rc::get_mut(&mut self.inner).unwrap().tracing =
+                                Some(hl_ptr as *const ());
+                            loc.unlock();
+                            return None;
+                        }
+                        Err(x) => {
+                            if x.is_locked() {
+                                // We probably raced with another thread locking this Location in order to
+                                // start tracing. It's unlikely to be worth us spending time contending
+                                // with that other thread.
+                                unsafe { Box::from_raw(hl_ptr) };
                                 return None;
                             }
-                            _ => unreachable!(),
+                            ls = x;
                         }
-                        yield_now();
-                        lp = loc.load(Ordering::Relaxed);
                     }
                 }
-
-                let loc_tid = unsafe { Arc::from_raw(lp.pointer_data::<ThreadIdInner>()) };
-                if let Some((_, ref tid)) = self.inner.tracer {
-                    // This thread is tracing something...
-                    if !Arc::ptr_eq(tid, &loc_tid) {
-                        // ...but we didn't start at the current Location.
-                        loc.store(lp, Ordering::Relaxed);
-                        mem::forget(loc_tid);
+            }
+        } else {
```

Comment:
This `else` is so far away from it's `if` that it makes it hard to find, especially with all the other nested conditionals in close proximity. Does it make sense to put the bodies of the arms into inlined functions? I think it does.

Reply:
I agree, this isn't ideal. However, I think we need to make some other changes in subsequent PRs that *might* negate the need for this -- I'm not really sure. Summary: I think we should wait a bit before trying to make this prettier.

Reply:
I'm OK with that.